python_modules/dagster/dagster/core/asset_defs/decorators.py

```diff
@@ -309,6 +313,154 @@ def inner(fn: Callable[..., Any]) -> AssetsDefinition:
+        )
+        asset_key_by_output_name = {
+            output_name: asset_key for asset_key, output_name in output_names_by_asset_key.items()
+        }
+        transformed_internal_asset_deps = {}
+        if internal_asset_deps:
+            for output_name, asset_keys in internal_asset_deps.items():
+                check.invariant(
+                    output_name in asset_key_by_output_name,
+                    f"output_name {output_name} specified in internal_asset_deps does not exist in the decorated function",
+                )
+                transformed_internal_asset_deps[asset_key_by_output_name[output_name]] = asset_keys
+
+        return AssetsDefinition(
+            input_names_by_asset_key=_infer_input_names_by_asset_key(
+                op_def,
+                asset_keys_by_input_name or {},
+            ),
+            output_names_by_asset_key=output_names_by_asset_key,
+            op=op_def,
+            asset_deps=transformed_internal_asset_deps or None,
+        )
+
+    return inner
+
+
+def _get_input_param_names(fn_params: List[funcsigs.Parameter]) -> List[str]:
+    is_context_provided = len(fn_params) > 0 and fn_params[0].name in get_valid_name_permutations(
+        "context"
+    )
+    return [
+        input_param.name for input_param in (fn_params[1:] if is_context_provided else fn_params)
+    ]
+
+
+def _infer_input_names_by_asset_key(
+    op_def: OpDefinition, asset_keys_by_input_name: Mapping[str, AssetKey]
+) -> Mapping[AssetKey, str]:
+    # Infer non-argument deps for inputs with type In(nothing) with AssetKey(input_name)
+
+    input_param_names = []
+    if isinstance(op_def.compute_fn, DecoratedSolidFunction):
+        params = get_function_params(op_def.compute_fn.decorated_fn)
+        input_param_names = _get_input_param_names(params)
+
+    for in_key in asset_keys_by_input_name.keys():
+        if in_key not in input_param_names:
+            raise DagsterInvalidDefinitionError(
+                f"Key '{in_key}' in provided asset_keys_by_input_name dict does not correspond to "
+                "any key provided in the ins dictionary of the decorated op or any argument "
+                "to the decorated function"
+            )
+
+    all_input_names = set(input_param_names) | op_def.ins.keys()
+    # If asset key is not supplied in asset_keys_by_input_name, create asset key
+    # from input name
+    inferred_input_names_by_asset_key: Dict[AssetKey, str] = {
+        asset_keys_by_input_name.get(input_name, AssetKey([input_name])): input_name
+        for input_name in all_input_names
+    }
+
+    return inferred_input_names_by_asset_key
+
+
+def _infer_output_names_by_asset_key(
+    op_def: OpDefinition, asset_keys_by_output_name: Mapping[str, AssetKey]
+) -> Mapping[AssetKey, str]:
+    inferred_output_name_by_asset_key: Dict[AssetKey, str] = {
+        asset_key: output_name for output_name, asset_key in asset_keys_by_output_name.items()
```

Comment:
I think for the case that there's only a single output named "result", we should use the name of the op as the output key rather than the name of the output.